Fix Open API doc for DateTime requirements

raziel057 opened the pull request:
When using the following requierement:

```
@Rest\QueryParam(name="updatedFrom", strict=true, nullable=true, requirements=@Constraints\DateTime("Y-m-d\TH:i:s"))
```

According to the OpenAPI 3.0 documentation https://swagger.io/specification/, 

We should have ``"format": "date-time"`` rather than ``"format": "DateTime"`` in the generated doc

GuilhemN commented:
If we go this way, maybe we should somehow merge the logic with https://github.com/nelmio/NelmioApiDocBundle/blob/master/ModelDescriber/Annotations/SymfonyConstraintAnnotationReader.php, what do you think?

phansys reviewed (commented):
Could you please add a test covering this fix?

phansys commented on RouteDescriber/FosRestDescriber.php:
```diff
@@ -104,6 +105,16 @@ private function getPattern($requirements)
+                // As defined per RFC3339
+                if ('Y-m-d\TH:i:s' === $requirements->format) {
+                    return 'date-time';
+                } elseif ('Y-m-d' === $requirements->format) {
+                    return 'date';
+                }
```
```suggestion
                }
                
                if ('Y-m-d' === $requirements->format) {
                    return 'date';
                }
```

raziel057 commented:
@phansys I just added the unit tests and fixed the CS

phansys commented on Tests/Functional/Controller/FOSRestController.php:
```diff
@@ -28,6 +29,8 @@ class FOSRestController
      * @RequestParam(name="Barraa", key="bar", requirements="\d+")
      * @RequestParam(name="baz", requirements=@IsTrue)
+     * @RequestParam(name="datetime", requirements=@DateTime("Y-m-d\TH:i:s"))
+     * @RequestParam(name="date", requirements=@DateTime("Y-m-d"))
```
Could you please add test cases for other valid PHP date formats, like `"c"`?

raziel057 commented:
@phansys  I just added support for 'c' format

phansys commented:
> @phansys I just added support for 'c' format

I don't know if we should support these formats (`"c"`, `\DateTimeInterface::ATOM`, etc).
Mainly, my suggestion is about covering the case where the user uses other formats which are valid in PHP, but exposing that not all of them are valid for the OAS schema.
BTW, I'd like also be consistent with the behavior around this between describers.

raziel057 commented:
Maybe I'm wrong but for me if the provided format doesn't match the OA spec we should not set the ``date`` or ``date-time`` format.

raziel057 commented:
> 
> 
> If we go this way, maybe we should somehow merge the logic with https://github.com/nelmio/NelmioApiDocBundle/blob/master/ModelDescriber/Annotations/SymfonyConstraintAnnotationReader.php, what do you think?

For me this point is not directly related and can be worked on another PR.

raziel057 commented:
When do you think this PR could be merged?

phansys commented:
>Maybe I'm wrong but for me if the provided format doesn't match the OA spec we should not set the date or date-time format.

I agree with that, is what I tried to say here:
>I don't know if we should support these formats ("c", \DateTimeInterface::ATOM, etc).
Mainly, my suggestion is about covering the case where the user uses other formats which are valid in PHP, but exposing that not all of them are valid for the OAS schema.

It would be nice to have test cases for scenarios where invalid OAS date formats (but valid for PHP) are passed, reflecting how the serialization behaves.

raziel057 commented:
@phansys I just fixed the format that should match according to the RFC3339 and added a new Assert for date with format that is not matching the RFC3339. In that case rather than generating a ``format: DateTime`` we doesn't set any format. At first I had in mind to generate a pattern given the date format but it seems overkill.

phansys commented on RouteDescriber/FosRestDescriber.php:
```diff
@@ -104,6 +105,17 @@ private function getPattern($requirements)
         if ($requirements instanceof Constraint && !$requirements instanceof Regex) {
+            if ($requirements instanceof DateTime) {
+                // As defined per RFC3339
+                if (\DateTime::RFC3339 === $requirements->format || 'c' === $requirements->format) {
```
```suggestion
                if (\DateTimeInterface::RFC3339 === $requirements->format || 'c' === $requirements->format) {
```

raziel057 replied:
\DateTimeInterface::RFC3339 is not supported by PHP 7.1 as you can see here:
https://github.com/nelmio/NelmioApiDocBundle/runs/2186348084

That's why I done the following change here:
https://github.com/nelmio/NelmioApiDocBundle/pull/1788/commits/d6a2d1240a853262a8005f6e4d8533127e1101f7

phansys commented on RouteDescriber/FosRestDescriber.php:
```diff
@@ -104,6 +105,17 @@ private function getPattern($requirements)
+                // As defined per RFC3339
+                if (\DateTime::RFC3339 === $requirements->format || 'c' === $requirements->format) {
+                    return 'date-time';
+                } elseif ('Y-m-d' === $requirements->format) {
+                    return 'date';
+                } else {
+                    return null;
+                }
```
```suggestion
                }
                
                if ('Y-m-d' === $requirements->format) {
                    return 'date';
                }

                return null;
```
BTW, I think we should keep the existing behavior if the provided format is not supported or throw an exception in that case.

raziel057 replied:
Whether we add a ``format: DateTime`` or not at all in the resulting OA doc, the Rest Client Generator (.net) will create a string parameter as the DateTime format is unknown. So for me it should not appear as not supported by https://swagger.io/specification/. I'm not for throwing an exception as if someone wants to provide an other PHP supported datetime format it should be possible but just ignored by Open API.

GuilhemN replied:
Throwing an exception would be too bad for DX imo, ignoring the constraint in case we don't know what to do looks like a fine compromise to me.

GuilhemN commented:
Thank you @raziel057 and @phansys for you review!
